File: src/services/sandbox.ts
/**
 * Sandbox service for code execution
 * 
 * Provides frontend interface to the Rust sandbox backend
 */

import { invoke } from '@tauri-apps/api/core';

/** Supported sandbox languages */
export type SandboxLanguage = 'python' | 'javascript' | 'shell';

/** Execution status */
export type ExecutionStatus = 
  | 'pending'
  | 'running'
  | 'completed'
  | 'failed'
  | 'timeout'
  | 'cancelled';

/** Sandbox permissions */
export interface SandboxPermissions {
  readFiles: boolean;
  writeFiles: boolean;
  networkAccess: boolean;
  maxExecutionMs: number;
  maxMemoryMb: number;
}

/** Default permissions for sandbox execution */
export const DEFAULT_PERMISSIONS: SandboxPermissions = {
  readFiles: true,
  writeFiles: false,
  networkAccess: false,
  maxExecutionMs: 30000,
  maxMemoryMb: 256,
};

/** Execution result from the sandbox */
export interface ExecutionResult {
  id: string;
  status: ExecutionStatus;
  stdout: string;
  stderr: string;
  exitCode: number | null;
  executionMs: number;
}

/** Runtime information */
export interface RuntimeInfo {
  language: string;
  version: string;
  available: boolean;
  path: string | null;
}

/**
 * Execute code in the sandbox
 */
export async function executeCode(
  code: string,
  language: SandboxLanguage,
  options?: {
    permissions?: Partial<SandboxPermissions>;
    workingDir?: string;
  }
): Promise<ExecutionResult> {
  const permissions = options?.permissions
    ? { ...DEFAULT_PERMISSIONS, ...options.permissions }
    : DEFAULT_PERMISSIONS;

  return invoke<ExecutionResult>('execute_code', {
    code,
    language,
    permissions,
    workingDir: options?.workingDir ?? null,
  });
}

/**
 * Cancel an active execution
 */
export async function cancelExecution(executionId: string): Promise<void> {
  return invoke('cancel_execution', { executionId });
}

/**
 * List available runtimes
 */
export async function listAvailableRuntimes(): Promise<RuntimeInfo[]> {
  return invoke<RuntimeInfo[]>('list_available_runtimes');
}

/**
 * Check if a specific runtime is available
 */
export async function isRuntimeAvailable(language: SandboxLanguage): Promise<boolean> {
  const runtimes = await listAvailableRuntimes();
  const runtime = runtimes.find(r => r.language === language);
  return runtime?.available ?? false;
}

/**
 * Get runtime info for a specific language
 */
export async function getRuntimeInfo(language: SandboxLanguage): Promise<RuntimeInfo | null> {
  const runtimes = await listAvailableRuntimes();
  return runtimes.find(r => r.language === language) ?? null;
}

File: src-tauri/src/commands/sandbox.rs
//! Sandbox Tauri commands
//!
//! Exposes sandbox functionality to the frontend

use crate::sandbox::{
    ExecutionRequest, ExecutionResult, RuntimeInfo, SandboxLanguage, SandboxPermissions,
    runtime::Runtime, security::SecurityValidator,
};
use std::path::PathBuf;
use std::sync::Arc;
use tauri::State;
use tokio::sync::Mutex;

/// Sandbox state managed by Tauri
pub struct SandboxState {
    pub runtime: Arc<Mutex<Runtime>>,
}

impl Default for SandboxState {
    fn default() -> Self {
        Self {
            runtime: Arc::new(Mutex::new(Runtime::new())),
        }
    }
}

/// Execute code in the sandbox
#[tauri::command]
pub async fn execute_code(
    state: State<'_, SandboxState>,
    code: String,
    language: String,
    permissions: Option<SandboxPermissions>,
    working_dir: Option<String>,
) -> Result<ExecutionResult, String> {
    // Parse language
    let lang = match language.to_lowercase().as_str() {
        "python" => SandboxLanguage::Python,
        "javascript" | "js" => SandboxLanguage::JavaScript,
        "shell" | "sh" | "bash" => SandboxLanguage::Shell,
        _ => return Err(format!("Unsupported language: {}", language)),
    };
    
    // Use default permissions if not provided
    let perms = permissions.unwrap_or_default();
    
    // Validate permissions
    SecurityValidator::validate_permissions(&perms)?;
    
    // Validate code for security issues
    SecurityValidator::validate_code(&code, &language)?;
    
    // Build execution request
    let request = ExecutionRequest {
        code,
        language: lang,
        permissions: perms,
        working_dir: working_dir.map(PathBuf::from),
    };
    
    // Execute
    let runtime = state.runtime.lock().await;
    runtime.execute(request).await
}

/// Cancel an active execution
#[tauri::command]
pub async fn cancel_execution(
    state: State<'_, SandboxState>,
    execution_id: String,
) -> Result<(), String> {
    let runtime = state.runtime.lock().await;
    runtime.cancel(&execution_id).await
}

/// List available runtimes
#[tauri::command]
pub async fn list_available_runtimes(
    state: State<'_, SandboxState>,
) -> Result<Vec<RuntimeInfo>, String> {
    let runtime = state.runtime.lock().await;
    Ok(runtime.list_available_runtimes())
}
